docs/design/K2_BASIC_CONCEPTS.md

```diff
@@ -0,0 +1,175 @@
+};
+
+struct Schema {
+    String name;
+    uint32_t version = 0;
+    std::vector<SchemaField> fields;
+
+    // All key fields must come before all value fields (by index), so that a key can be
+    // constructed for a read request without knowing the schema version
+    std::vector<uint32_t> partitionKeyFields;
+    std::vector<uint32_t> rangeKeyFields;
+};
+```
+
+
+## Quick overview
+This is an overview of how data is organized in K2. Details on each part are following
+
+![Data Deployment](../images/Data_Deployment.png)
+
+## Cluster, nodes, nodepool
+A recap from the K2PROCESS.md document:
+A cluster is a set of machines used to deploy data which is managed as a single unit. Wd do not have a notion of cluster in code - it is just how we may refer in documentation to a set of machines which are used to deploy a `Collection`(see below). The data lives in partitions, each deployed to a single CPU Core with some particular fixed memory available to it (4-12GB). We call this a `node` and is what messages in the system are send to. We start these on a given machine as individual threads in a process which we call the `nodepool`. Currently we deploy nodepools one per CPU socket, so a 2-socket machine (each socket with say 20cores) would get 2 nodepools, each running 20 threads for a total of 40 nodes available for deployment on this machine.
+
+## Collection
+Collections in K2 is how data is deployed on a set of nodes. Thus to work with data, a collection must already exist:
+
+![collections](../images/K2Collection.png)
+
+A collection is akin to a `Database` in the SQL world - it allows the data owner to manage the scalability and capacity of a bunch of datasets(related or not) together. For example, this could be the datasets for an online business: {sellers, users, orders, items, listings, offers, ...}. As the business grows, all of these datasets would see growth and it makes sense to manage their capacity and request/storage distribution as a single unit.
+
+It should now be clear why we require the `collectionName` in each request - the requests are always collection-specific. In addition, we support multiple collections in a single transaction so that the users can still transact over data which isn't necessarily all in a single collection.
+
+## Schema
+A Schema is what we use to describe the data in a data record. In particular, it specifies
+- the names and types for all fields
+- which fields comprise the partitionKey
+- which fields comprise the rangeKey
+
+Schemas have names and are versioned, and is akin to a `table` in the SQL world. All of the data in a particular schema is supposed to be part of a single data set (e.g. `Users`, or `Orders`). Via versioning, the users can perform schema migration, though we do not support changes to the key fields (partition/range).
+
+- Schemas are created and are therefore scoped inside a particular collection.
+- When reading data, the user has to specify from which schema they are reading
```

Comment:
Could link to the schema design docs here if you want.

Reply:
linked